Add kube-proxy non-privileged blog post (#43680)
kubernetes/website: content/en/blog/_posts/0000-00-00-kube-proxy-non-privileged.md

---
layout: blog

> title: "Non-privileged kube-proxy container"

Reviewer: ```suggestion
title: "Kubernetes 1.29: Support for Running kube-proxy in a Non-privileged Container"
```

date: 0000-00-00
slug: kube-proxy-non-privileged
---

**Author**: Lars Ekman

> This post describes how the `--init-only` flag added in K8s v1.29 can

Reviewer: ```suggestion
This post describes how the `--init-only` flag to `kube-proxy`, added in K8s v1.29, can
```

be used to perform configuration that requires privileged mode in an
initContainer, while the main `kube-proxy` container may run with a

> stricter securityContext.

Reviewer: stricter `securityContext`.

> stricter securityContext.

Reviewer 2: maybe flip that around to put the cool part first: "can be used to run the main kube-proxy container in a stricter securityContext by performing the configuration that requires privileged mode in a separate initContainer"

Author: I also removed the "added in K8s v1.29", since it's now in the title.

> Please note that `kube-proxy` can be installed in different ways, but
> installers that installs `kube-proxy` as a POD, like `kubeadm`, can
> now use this function for improved security.

Reviewer: ```suggestion
Please note that `kube-proxy` can be installed in different ways. For
installers such as `kubeadm` that install `kube-proxy` to run as a Pod, the new
functionality can help you improve security.
```

> Please note that `kube-proxy` can be installed in different ways. For
> installers such as `kubeadm` that install `kube-proxy` to run as a Pod, the new
> functionality can help you improve security.

Reviewer 2: This can help you improve security if it _doesn't_ run as a pod too...

```suggestion
Please note that `kube-proxy` can be installed in different ways. The
examples below assume that kube-proxy is run from a pod, but similar
changes could be made in clusters where it is run as a system service.
```

## Background

It is undesirable to run a server container like `kube-proxy` in
privileged mode. Security aware users wants to [use capabilities instead](

> https://github.com/kubernetes/kubernetes/issues/112171).

Reviewer: Readers will expect this to take them to a guide, not an issue. I'd omit the link.

If `kube-proxy` is installed as a POD, the initialization requires
"privileged" mode, mostly for setting sysctl's. However, for some time
now `kube-proxy` doesn't alter sysctl's and other settings that
already have the desired value. In theory this can be used to do all

> privileged configuration in an initContainer.

Reviewer: ```suggestion
privileged configuration in an [init container](/docs/concepts/workloads/pods/init-containers/).
```

> If `kube-proxy` is installed as a POD, the initialization requires
> "privileged" mode, mostly for setting sysctl's. However, for some time
> now `kube-proxy` doesn't alter sysctl's and other settings that
> already have the desired value. In theory this can be used to do all
> privileged configuration in an [init container](/docs/concepts/workloads/pods/init-containers/).

Reviewer 2: I think this would be confusing to someone who doesn't already understand the problem...

```suggestion
If `kube-proxy` is installed as a POD, the initialization requires
"privileged" mode, mostly for setting sysctl's. However, `kube-proxy`
only tries to set the sysctl's if they don't already have the right
values. In theory, then, if a privileged [init container](/docs/concepts/workloads/pods/init-containers/)
set the sysctls to the right values, then `kube-proxy` could run
unprivileged.
```

The problem is to know *what* to setup. Until now the only option has
been to read the source, but with `--init-only` the setup is done
*exactly* as on a normal start, and then `kube-proxy` exits.

> ## Kube-proxy initialization in an initContainer

Reviewer: ```suggestion
## Initializing kube-proxy in an init container
```

> The problem is to know *what* to setup. Until now the only option has
> been to read the source, but with `--init-only` the setup is done
> *exactly* as on a normal start, and then `kube-proxy` exits.

Reviewer 2: ```suggestion
The problem is to know *what* to setup. Until now the only option has
been to read the source to see what changes `kube-proxy` would have
made, but with `--init-only` you can have `kube-proxy` itself do the setup
*exactly* as on a normal start, and then exit.
```

> Traditionally `kube-proxy` runs in "privileged mode (manifests
> narrowed):

Reviewer: ```suggestion
Usually, cluster operators run `kube-proxy` in a privileged security context. Here's a snippet of an
example manifest:
```

```yaml
apiVersion: apps/v1
kind: DaemonSet
metadata:
  labels:
    k8s-app: kube-proxy
spec:
  template:
    spec:
      containers:
      - name: kube-proxy
        command:
        - /usr/local/bin/kube-proxy
        - --config=/var/lib/kube-proxy/config.conf
        - --hostname-override=$(NODE_NAME)
        securityContext:
          privileged: true
```

But now it is possible to use:

```yaml
apiVersion: apps/v1
kind: DaemonSet
metadata:
  labels:
    k8s-app: kube-proxy
spec:
  template:
    spec:
      initContainers:
      - name: kube-proxy-init
        command:
        - /usr/local/bin/kube-proxy
        - --config=/var/lib/kube-proxy/config.conf

>         - --init-only

Reviewer 2: I would include the `--hostname-override` here as well. _We_ know that it doesn't matter, but end users won't have a good sense of which options might matter and which ones don't, so they should just specify all the same options they specify for the real container.

        securityContext:
          privileged: true
      containers:
      - name: kube-proxy
        command:
        - /usr/local/bin/kube-proxy
        - --config=/var/lib/kube-proxy/config.conf
        - --hostname-override=$(NODE_NAME)
        securityContext:
          capabilities:
            add: ["NET_ADMIN"]
```

## Improve security even more

> 

Reviewer 2: ???

Author: Ah, this was where I intended to describe "User Namespaces", but they can't be used. I will remove the header

## Summary

The `--init-only` flag can be used to perform privileged
initialization in an initContainer and run the main container with

> "NET_ADMIN" capabilities only. Installers like `kubeadm` may be

Reviewer: ```suggestion
`NET_ADMIN` capabilities only. Installers like `kubeadm` may be
```

> altered to use this feature.

Reviewer 2: That's very passive... you should either submit a patch to make kubeadm do it, or if not (or if there's not time to get it in for 1.29), at least file a bug against it suggesting they should, and then you say "kubeadm will like be altered to use this feature in the future".

Author: Will do that tomorrow

only tries to set the sysctl's if they don't already have the right
values. In theory, then, if a privileged [init container](/docs/concepts/workloads/pods/init-containers/)
set the sysctls to the right values, then `kube-proxy` could run
unprivileged.

The problem is to know *what* to setup. Until now the only option has
been to read the source to see what changes `kube-proxy` would have
made, but with `--init-only` you can have `kube-proxy` itself do the setup
*exactly* as on a normal start, and then exit.

## Initializing kube-proxy in an init container

Usually, cluster operators run `kube-proxy` in a privileged security context. Here's a snippet of an
example manifest:

```yaml
apiVersion: apps/v1
kind: DaemonSet
metadata:
  labels:
    k8s-app: kube-proxy
spec:
  template:
    spec:
      containers:
      - name: kube-proxy
        command:
        - /usr/local/bin/kube-proxy
        - --config=/var/lib/kube-proxy/config.conf
        - --hostname-override=$(NODE_NAME)
        securityContext:
          privileged: true
```

But now it is possible to use:

```yaml
apiVersion: apps/v1
kind: DaemonSet
metadata:
  labels:
    k8s-app: kube-proxy
spec:
  template:

>     spec:

Reviewer 3: best to include:
```
image: registry.k8s.io/kube-proxy:v1.28.2
magePullPolicy: IfNotPresent
```

for both init and regular containers in the example.

"privileged" mode, mostly for setting sysctl's. However, `kube-proxy`
only tries to set the sysctl's if they don't already have the right
values. In theory, then, if a privileged [init container](/docs/concepts/workloads/pods/init-containers/)
set the sysctls to the right values, then `kube-proxy` could run
unprivileged.

The problem is to know *what* to setup. Until now the only option has
been to read the source to see what changes `kube-proxy` would have
made, but with `--init-only` you can have `kube-proxy` itself do the setup
*exactly* as on a normal start, and then exit.

## Initializing kube-proxy in an init container

Usually, cluster operators run `kube-proxy` in a privileged security context. Here's a snippet of an
example manifest:

```yaml
apiVersion: apps/v1
kind: DaemonSet
metadata:
  labels:
    k8s-app: kube-proxy
spec:
  template:
    spec:
      containers:
      - name: kube-proxy
        command:
        - /usr/local/bin/kube-proxy
        - --config=/var/lib/kube-proxy/config.conf
        - --hostname-override=$(NODE_NAME)
        securityContext:
          privileged: true
```

But now it is possible to use:

```yaml

> apiVersion: apps/v1
> kind: DaemonSet
> metadata:
>   labels:
>     k8s-app: kube-proxy
> spec:
>   template:
>     spec:
>       initContainers:
>       - name: kube-proxy-init
>         command:
>         - /usr/local/bin/kube-proxy
>         - --config=/var/lib/kube-proxy/config.conf
>         - --hostname-override=$(NODE_NAME)
>         - --init-only
>         securityContext:
>           privileged: true
>       containers:
>       - name: kube-proxy
>         command:
>         - /usr/local/bin/kube-proxy
>         - --config=/var/lib/kube-proxy/config.conf
>         - --hostname-override=$(NODE_NAME)
>         securityContext:
>           capabilities:
>             add: ["NET_ADMIN"]

Reviewer: Try this, which I think is a complete manifest.
```suggestion
apiVersion: apps/v1
kind: DaemonSet
metadata:
  namespace: kube-system
  name: kube-proxy-linux
  labels:
    k8s-app: kube-proxy
spec:
  selector:
    matchLabels:
      k8s-app: kube-proxy
      kubernetes.io/os: linux
  template:
    metadata:
      labels:
        k8s-app: kube-proxy
        kubernetes.io/os: linux
      annotations:
        kubernetes.io/description: >-
          Node-level network proxy; part of Kubernetes
    spec:
      tolerations:
      - key: node-role.kubernetes.io/control-plane
        effect: NoSchedule
        operator: Exists
      os: { name: linux } # On Windows, kube-proxy needs to run privileged
                        # You will need a second DaemonSet if your cluster
                        # includes Windows nodes
      nodeSelector:
        kubernetes.io/os: linux
      initContainers:
      - name: kube-proxy-init
        image: registry.k8s.io/kube-proxy:v1.28.2 # adjust this to the Kubernetes version you use
        imagePullPolicy: IfNotPresent
        # you should set resource requests here
        # see https://k8s.io/docs/concepts/configuration/manage-resources-containers/
        command:
        - /usr/local/bin/kube-proxy
        - --config=/var/lib/kube-proxy/config.conf
        - --hostname-override=$(NODE_NAME)
        - --init-only
        securityContext:
          privileged: true
      containers:
      - name: kube-proxy
        image: registry.k8s.io/kube-proxy:v1.28.2 # adjust this to the Kubernetes version you use
        imagePullPolicy: IfNotPresent
        # you should set resource requests here
        # see https://k8s.io/docs/concepts/configuration/manage-resources-containers/
        command:
        - /usr/local/bin/kube-proxy
        - --config=/var/lib/kube-proxy/config.conf
        - --hostname-override=$(NODE_NAME)
        securityContext:
          capabilities:
            add: ["NET_ADMIN"]
    terminationGracePeriodSeconds: 30
```

Reviewer: I guess this doesn't cover a ConfigMap though, and maybe we should talk about that.

Author: I disagree on this. The kubeadm maintainers, and others who really will use this, are totally capable of figuring out the exact procedure. IMHO the blog should give normal users an _idea_ on what is done, not be a complete instruction. It should be TL;DR; Besides, the kubeadm manifest is likely not the only one used

Author: The confirMap is not affected. `--init-only` is a CLI option only, it can't be configured in the config file

Reviewer: OK, but: every bit we miss out, let's clearly tell readers what's not there. We don't need to give all the details but let's make it **very very** obvious that this manifest is incomplete.

Even if we do all that, the website repo is likely to get issues filed by people who don't get it. Let's do what we reasonably can to minimize the rate of those issues.

Author: Updated.

It might be even more issues if we give the _illusion_ that the manifests are complete, and are compatible with the ones the users may have (which may not be kubeadm generated).

Reviewer: So, on that: let's add a very clear note that this still isn't complete: nothing populates `/var/lib/kube-proxy/config.conf`.

```

## Summary

The `--init-only` flag can be used to perform privileged

> initialization in an initContainer and run the main container with

Reviewer: ```suggestion
initialization in an init container and run the main container with
```

## Initializing kube-proxy in an init container

*The example manifests below are not complete, but narrowed down to what is

> essential to illustrate the function.*

Reviewer: Thanks, but I think we can safely assume that a good proportion of readers' eyes slip right past this and start at the manifest.

Reviewer: (we should keep the note here, but we should also make it clear in the example manifest about which details aren't included)
